crates/pipeline-manager/src/compiler/rust_compiler.rs

```diff
@@ -1648,6 +1864,32 @@ async fn cleanup_rust_compilation(
         )
         .collect();
 
+    let valid_program_info_filenames: Vec<String> = existing_pipeline_programs
```

Comment:
regarding cleanup:

when is entry added in existing_pipeline_programs?

what we don't want is, say if we are uploading program_info file, but it wasn't part of this vec, and cleanup gets executed, deleting that file.

^ basically similar to race condition we faced with binary upload

Reply:
That's a good point. I implemented this code by analogy with how we cleanup binary files, so whatever correctness argument applies there should hold. However, I am not sure what prevents the race condition you're describing. @snkas , do you know the answer?